lta/rest_server.py

```diff
@@ -269,79 +278,108 @@ class FilesActionsPopHandler(BaseLTAHandler):
+            db_files.append(row)
+        # do a sanity check; individual files aren't too big
+        for db_file in db_files:
+            db_file_size = db_file["catalog"]["file_size"]
+            db_file_uuid = db_file["uuid"]
+            if db_file_size > bundle_size:
+                # TODO: Move the monster file to quarantine?
+                raise tornado.web.HTTPError(400, reason=f"cannot bundle file {db_file_uuid} (size: {db_file_size}); bundle size is {bundle_size}")
+        # do a sanity check; enough files to make a full bundle
+        full_house = (len(db_files) >= limit)
+        full_size = 0
+        for db_file in db_files:
+            full_size = full_size + db_file["catalog"]["file_size"]
+        if (full_house) and (full_size < bundle_size):
+            raise tornado.web.HTTPError(400, reason="limit must be raised to reach full bundle size")
+        if (not force) and (full_size < bundle_size):
             self.write({'results': []})
             return
-        # pack those files into bundle sized bins
-        bins = binpacking.to_constant_volume(src_tuples, bundle_size, weight_pos=1)
-        bin_count = len(bins)-1
-        # if we got exactly one bin
-        if bin_count == 0:
-            # bail, if we're not forcing it
-            if not force:
-                self.write({'results': []})
-                return
-            # otherwise, pikachu, I choose you
-            bin_choice = 0
-        # otherwise, we have multiple bins to choose from
-        else:
-            # randomly select one of the available bins
-            bin_choice = floor(bin_count * random())
-        # get a list of the files in the bin
-        pop_bin = bins[bin_choice]
-        bin_files = [self.db["Files"][x[0]] for x in pop_bin]
+        # use The Price is Right rules to pick files for bundling
+        build_files = []
+        build_size = 0
+        for db_file in db_files:
+            db_file_size = db_file["catalog"]["file_size"]
+            if (build_size + db_file_size) > bundle_size:
+                break
+            build_files.append(db_file)
+            build_size = build_size + db_file_size
         # update the status of the files we're handing out
         claim_time = now()
-        for bin_file in bin_files:
-            bin_file["claimant"] = pop_body
-            bin_file["claimed"] = True
-            bin_file["claim_time"] = claim_time
-            bin_file["status"] = "processing"
+        for build_file in build_files:
+            build_file["claimant"] = pop_body
+            build_file["claimed"] = True
+            build_file["claim_time"] = claim_time
+            build_file["status"] = "processing"
+            update_query = {"uuid": build_file["uuid"]}
+            update_doc = {"$set": build_file}
+            ret = await sdf.find_one_and_update(update_query, update_doc)
+            if not ret:
+                raise tornado.web.HTTPError(500, reason=f"unable to claim file for bundling")
```

Comment:
This shouldn't be a fatal error.  We need to allow two pop operations to happen in parallel, and they might try to step on each other.